fingerprint2.js

```diff
@@ -316,181 +336,188 @@
         done(keys)
       })
     },
-    // kudos to http://www.lalit.org/lab/javascript-css-font-detect/
-    jsFontsKey: function (keys, done) {
+    jsFontsKey: function(keys, done, reject) {
       var that = this
       // doing js fonts detection in a pseudo-async fashion
       return setTimeout(function () {
-        // a font will be compared against all the three default fonts.
-        // and if it doesn't match all 3 then that font is not available.
-        var baseFonts = ['monospace', 'sans-serif', 'serif']
+        try {
+          that.jsFontsKeySynchronous(keys, done, that)
+        } catch(e) {
+          reject(e)
+        }
+      }, 1)
+    },
+    // kudos to http://www.lalit.org/lab/javascript-css-font-detect/
+    jsFontsKeySynchronous: function (keys, done, that) {
```

Comment:
Can we make this function behave like any other synchronous component function, i.e. take `keys` and return `keys`? Then we could have all the promise/async related code in `jsFontsKey`.

Reply:
I do not know why you use the setTimeout function in the first place. If it was synchronous, everything would be much simpler.
I assume it is because of UI lag reason. The fonts part is probably computation-heavy and could be user unfriendly doing it synchronously.

Reply:
I guess so! I'm not the original author of this project so I'm not 100% sure about the reasoning here but I'm sure there's good reasons! :)